### ISSUE TITLE
Internal maximum timestamp values overflow on some 32 Bit systems

### ISSUE THREAD

**Issue by gsauthof** [ronf/asyncssh#163]
In the process of packaging asyncssh for Fedora I noticed that some tests fail due time_t overflows on certain 32 Bit Linux systems.

Example:

    Traceback (most recent call last):
      File "/builddir/build/BUILD/asyncssh-1.13.3/tests/test_x11.py", line 272, in setUpClass
	super().setUpClass()
      File "/builddir/build/BUILD/asyncssh-1.13.3/tests/util.py", line 348, in setUpClass
	cls.loop.run_until_complete(cls.asyncSetUpClass())
      File "/usr/lib/python3.7/asyncio/base_events.py", line 568, in run_until_complete
	return future.result()
      File "/builddir/build/BUILD/asyncssh-1.13.3/tests/server.py", line 124, in asyncSetUpClass
	ckey_ecdsa, 'OU=name', principals=['ckey'])
      File "/builddir/build/BUILD/asyncssh-1.13.3/asyncssh/public_key.py", line 620, in generate_x509_user_certificate
	hash_alg, comment)
      File "/builddir/build/BUILD/asyncssh-1.13.3/asyncssh/public_key.py", line 229, in _generate_x509_certificate
	hash_alg, comment)
      File "/builddir/build/BUILD/asyncssh-1.13.3/asyncssh/public_key.py", line 1614, in generate
	hash_alg, comment)
      File "/builddir/build/BUILD/asyncssh-1.13.3/asyncssh/crypto/x509.py", line 302, in generate_x509_certificate
	builder = builder.not_valid_after(_to_generalized_time(valid_before))
      File "/builddir/build/BUILD/asyncssh-1.13.3/asyncssh/crypto/x509.py", line 59, in _to_generalized_time
	return datetime.utcfromtimestamp(max(1, min(t, _gen_time_max)))
    OverflowError: timestamp out of range for platform time_t

This is caused by `_gen_time_max` overflowing `time_t` - because it's set like this (under Linux):

    _gen_time_max = datetime.max.replace(tzinfo=timezone.utc).timestamp() - 1

In total this make many tests fail. There is some variation in the stack traces, but they all fail inside `_to_generalized_time()`.

I can think of 2 alternative ways how to fix this:

1. use a probing scheme when initializing _gen_time_max (e.g. for each ~ 2^63-1, 2^32-1, 2^31-1 use the largest one that doesn't overflow)
2. replace the internal unix-time representation with a timezone-aware datetime implementation

The first alternative is kind of messy  - since time_t might be signed or unsigned and might even be 64 Bit wide on 32 Bit systems. The [patch is relatively small](https://github.com/gsauthof/asyncssh/commit/1a6467a03b6546e7f164c68ac8161c0315cb254f), though. This is what I currently use for the Fedora package.

The second alternative mostly changes some internal implementation details. But arguably this should yield more robust code (i.e. more portable, easier to maintain). The [most interesting part of this change](https://github.com/gsauthof/asyncssh/commit/254b3cdc1248d211cc883b48c1e0d6dec47236c2) is in `_parse_time()` - that returns a datetime object instead of a unix-time value. The rest is mostly replacement of default values and the elimination of `_to_generalized_time()` and related  initialization.

**Comment by ronf** [comment]
Thanks very much for your report, and for the proposed code changes!

Since _parse_time is shared for both OpenSSH certificates and X.509 certificates and the SSH certificates store timestamps as a UInt64 value, I'd prefer not to change _parse_time() to return datetime values. I'd like to allow any legal 64-bit value to be used when creating OpenSSH certificates. However, I think a good alternative might be to not try and calculate a max epoch time in the X.509 code. Instead, that code can convert the provided epoch time into a datetime and simply check for an error. If it hits one, it can replace the value with datetime.max. Here's the patch I have in mind:

```patch
diff --git a/asyncssh/crypto/x509.py b/asyncssh/crypto/x509.py
index 8718472..42eee5e 100644
--- a/asyncssh/crypto/x509.py
+++ b/asyncssh/crypto/x509.py
@@ -45,18 +45,26 @@ _hashes = {h.name: h for h in (MD5, SHA1, SHA224, SHA256, SHA384, SHA512)}
 
 _nscomment_oid = x509.ObjectIdentifier('2.16.840.1.113730.1.13')
 
+_datetime_min = datetime.utcfromtimestamp(0).replace(microsecond=1,
+                                                     tzinfo=timezone.utc)
+
 if sys.platform == 'win32': # pragma: no cover
     # Windows' datetime.max is year 9999, but timestamps that large don't work
-    _gen_time_max = datetime(2999, 12, 31, 23, 59, 59, 999999,
-                             tzinfo=timezone.utc).timestamp() - 1
+    _datetime_max = datetime.max.replace(year=2999, tzinfo=timezone.utc)
 else:
-    _gen_time_max = datetime.max.replace(tzinfo=timezone.utc).timestamp() - 1
+    _datetime_max = datetime.max.replace(tzinfo=timezone.utc)
 
 
 def _to_generalized_time(t):
     """Convert a timestamp value to a datetime"""
 
-    return datetime.utcfromtimestamp(max(1, min(t, _gen_time_max)))
+    if t <= 0:
+        return _datetime_min
+    else:
+        try:
+            return datetime.utcfromtimestamp(t).replace(tzinfo=timezone.utc)
+        except (OSError, OverflowError):
+            return _datetime_max
 
 
 def _to_purpose_oids(purposes):
```

Could you give this a try and let me know if you still run into overflow problems?

The reason i check for OSError in addition to OverflowError is that some time values produce the following instead of the error you reported:

```
OSError: [Errno 84] Value too large to be stored in data type
```

For instance, I saw this when calling:

```python
datetime.utcfromtimestamp(0x7fffffffffffffff)
```

I was hoping I could use the approach above to eliminate the need to manually set the datetime max year to 2999 on Windows, but unfortunately the error which occurs in that case happens down inside the cryptography code. So, I still need to manually hard-code that. Here's the error I see when I don't do this:

```
Traceback (most recent call last):
  File "Z:\src\asyncssh\tests\test_x509.py", line 124, in test_valid_self
    cert = self.generate_certificate()
  File "Z:\src\asyncssh\tests\test_x509.py", line 55, in generate_certificate
    hash_alg, comment)
  File "Z:\src\asyncssh\asyncssh\crypto\x509.py", line 359, in generate_x509_certificate
    cert = builder.sign(signing_key.pyca_key, hash_alg, default_backend())
  File "C:\Python36\lib\site-packages\cryptography\x509\base.py", line 570, in sign
    return backend.create_x509_certificate(self, private_key, algorithm)
  File "C:\Python36\lib\site-packages\cryptography\hazmat\backends\openssl\backend.py", line 800, in create_x509_certificate
    self._raise_time_set_error()
  File "C:\Python36\lib\site-packages\cryptography\hazmat\backends\openssl\backend.py", line 842, in _raise_time_set_error
    "Invalid time. This error can occur if you set a time too far in "
ValueError: Invalid time. This error can occur if you set a time too far in the future on Windows.
```

**Comment by gsauthof** [comment]
I just tested my time-probing patch on that 32 Bit system. Not the datetime one. With that result I'm sure that it would fail in a similar way.

**Comment by ronf** [comment]
It's a bit ugly, but here's an attempt to work around the cryptography issue, while still allowing the larger datetime values to work on systems with a larger time_t:

```patch
diff --git a/asyncssh/crypto/x509.py b/asyncssh/crypto/x509.py
index 8718472..294d557 100644
--- a/asyncssh/crypto/x509.py
+++ b/asyncssh/crypto/x509.py
@@ -45,18 +45,33 @@ _hashes = {h.name: h for h in (MD5, SHA1, SHA224, SHA256, SHA384, SHA512)}
 
 _nscomment_oid = x509.ObjectIdentifier('2.16.840.1.113730.1.13')
 
+_datetime_min = datetime.utcfromtimestamp(0).replace(microsecond=1,
+                                                     tzinfo=timezone.utc)
+
+_datetime_32bit_max = datetime.utcfromtimestamp(2**31 - 1).replace(
+    tzinfo=timezone.utc)
+
 if sys.platform == 'win32': # pragma: no cover
     # Windows' datetime.max is year 9999, but timestamps that large don't work
-    _gen_time_max = datetime(2999, 12, 31, 23, 59, 59, 999999,
-                             tzinfo=timezone.utc).timestamp() - 1
+    _datetime_max = datetime.max.replace(year=2999, tzinfo=timezone.utc)
 else:
-    _gen_time_max = datetime.max.replace(tzinfo=timezone.utc).timestamp() - 1
+    _datetime_max = datetime.max.replace(tzinfo=timezone.utc)
 
 
 def _to_generalized_time(t):
     """Convert a timestamp value to a datetime"""
 
-    return datetime.utcfromtimestamp(max(1, min(t, _gen_time_max)))
+    if t <= 0:
+        return _datetime_min
+    else:
+        try:
+            return datetime.utcfromtimestamp(t).replace(tzinfo=timezone.utc)
+        except (OSError, OverflowError):
+            try:
+                _datetime_max.timestamp()
+                return _datetime_max
+            except (OSError, OverflowError):
+                return _datetime_32bit_max
 
 
 def _to_purpose_oids(purposes):
```

Would you mind giving this a try?

**Comment by gsauthof** [comment]
Sure, this also fails on that 32 Bit Linux machine, e.g.:

    ======================================================================
    ERROR: setUpClass (tests.test_logging._TestLogging)
    ----------------------------------------------------------------------
    Traceback (most recent call last):
      File "/builddir/build/BUILD/asyncssh-1.13.3/tests/util.py", line 348, in setUpClass
	cls.loop.run_until_complete(cls.asyncSetUpClass())
      File "/usr/lib/python3.7/asyncio/base_events.py", line 568, in run_until_complete
	return future.result()
      File "/builddir/build/BUILD/asyncssh-1.13.3/tests/server.py", line 124, in asyncSetUpClass
	ckey_ecdsa, 'OU=name', principals=['ckey'])
      File "/builddir/build/BUILD/asyncssh-1.13.3/asyncssh/public_key.py", line 620, in generate_x509_user_certificate
	hash_alg, comment)
      File "/builddir/build/BUILD/asyncssh-1.13.3/asyncssh/public_key.py", line 229, in _generate_x509_certificate
	hash_alg, comment)
      File "/builddir/build/BUILD/asyncssh-1.13.3/asyncssh/public_key.py", line 1614, in generate
	hash_alg, comment)
      File "/builddir/build/BUILD/asyncssh-1.13.3/asyncssh/crypto/x509.py", line 372, in generate_x509_certificate
	cert = builder.sign(signing_key.pyca_key, hash_alg, default_backend())
      File "/usr/lib/python3.7/site-packages/cryptography/x509/base.py", line 577, in sign
	return backend.create_x509_certificate(self, private_key, algorithm)
      File "/usr/lib/python3.7/site-packages/cryptography/hazmat/backends/openssl/backend.py", line 796, in create_x509_certificate
	calendar.timegm(builder._not_valid_after.timetuple())
    OverflowError: integer 253402300799 does not fit '32-bit int'

The reason is that the `timestamp()` call in your patch never throws. The only reliable test for the maximum `time_t` value is to again call `utcfromtimestamp()`. For example:

    datetime.utcfromtimestamp(_datetime_max.timestamp())

(cf. my probing patch)

According to the standard,s a 32 `time_t` might even be unsigned - thus, using 2**31-1 as maximum wouldn't use the additional headroom. But I can't find any documentation on real machines that use unsigned 32 Bit time_t. It's probably more likely that we see 64 Bit signed time_t on more 32 Bit systems as 2038 is coming closer.

I've checked the 32 Bit Fedora Linux system I'm testing against and it defines time_t as signed 32 Bit (`long int`).

**Comment by ronf** [comment]
> The reason is that the timestamp() call in your patch never throws. The only reliable test for the maximum time_t value is to again call utcfromtimestamp().

I was afraid that might be the case, but was hoping just calculating the timestamp() and have it overflow a time_t might have been enough.

> According to the standard, a 32 time_t might even be unsigned - thus, using 2**31-1 as maximum wouldn't use the additional headroom. But I can't find any documentation on real machines that use unsigned 32 Bit time_t. It's probably more likely that we see 64 Bit signed time_t on more 32 Bit systems as 2038 is coming closer.

Yeah - MacOS these days has a signed 64-bit time_t, so there's basically no issue with it representing any reasonable time value. It looks like it hits a limit only when the year exceeds a signed 32-bit value.

I realized when I was putting together the workaround that dropping all the way to signed 32-bit might be more than is needed on some systems, but it didn't seem worth the added complexity to try and address that. Hopefully all systems will update to a 64-bit time_t well before 2038.

Hopefully this is the last attempt needed at this -- thanks again for your support in testing it!

```patch
diff --git a/asyncssh/crypto/x509.py b/asyncssh/crypto/x509.py
index 8718472..0f66c9d 100644
--- a/asyncssh/crypto/x509.py
+++ b/asyncssh/crypto/x509.py
@@ -45,18 +45,35 @@ _hashes = {h.name: h for h in (MD5, SHA1, SHA224, SHA256, SHA384, SHA512)}
 
 _nscomment_oid = x509.ObjectIdentifier('2.16.840.1.113730.1.13')
 
+_datetime_min = datetime.utcfromtimestamp(0).replace(microsecond=1,
+                                                     tzinfo=timezone.utc)
+
+_datetime_32bit_max = datetime.utcfromtimestamp(2**31 - 1).replace(
+    tzinfo=timezone.utc)
+
 if sys.platform == 'win32': # pragma: no cover
     # Windows' datetime.max is year 9999, but timestamps that large don't work
-    _gen_time_max = datetime(2999, 12, 31, 23, 59, 59, 999999,
-                             tzinfo=timezone.utc).timestamp() - 1
+    _datetime_max = datetime.max.replace(year=2999, tzinfo=timezone.utc)
 else:
-    _gen_time_max = datetime.max.replace(tzinfo=timezone.utc).timestamp() - 1
+    _datetime_max = datetime.max.replace(tzinfo=timezone.utc)
 
 
 def _to_generalized_time(t):
     """Convert a timestamp value to a datetime"""
 
-    return datetime.utcfromtimestamp(max(1, min(t, _gen_time_max)))
+    if t <= 0:
+        return _datetime_min
+    else:
+        try:
+            return datetime.utcfromtimestamp(t).replace(tzinfo=timezone.utc)
+        except (OSError, OverflowError):
+            try:
+                # Work around a bug in cryptography which shows up on
+                # systems with a small time_t.
+                datetime.utcfromtimestamp(_datetime_max.timestamp() - 1)
+                return _datetime_max
+            except (OSError, OverflowError): # pragma: no cover
+                return _datetime_32bit_max
 
 
 def _to_purpose_oids(purposes):
```

**Comment by gsauthof** [comment]
Good news, with the latest revision of your patch all tests succeed on that 32 Bit machine (cf. https://kojipkgs.fedoraproject.org//work/tasks/6059/29316059/build.log).

**Comment by ronf** [comment]
Thanks so much for your help! This change is now commit 0fdea14 in the "develop" branch and will be included in the next release.

**Comment by ronf** [comment]
This is now available in AsyncSSH 1.14.0.
